serde-generate/src/typescript.rs

```diff
@@ -0,0 +1,747 @@
+        formats
+            .iter()
+            .map(|f| self.quote_type(f))
+            .collect::<Vec<_>>()
+            .join(sep)
+    }
+
+    fn output_helpers(&mut self, registry: &Registry) -> Result<()> {
+        let mut subtypes = BTreeMap::new();
+        for format in registry.values() {
+            format
+                .visit(&mut |f| {
+                    if Self::needs_helper(f) {
+                        subtypes.insert(common::mangle_type(f), f.clone());
+                    }
+                    Ok(())
+                })
+                .unwrap();
+        }
+
+        writeln!(self.out, "export class Helpers {{")?;
+        self.out.indent();
+        for (mangled_name, subtype) in &subtypes {
+            self.output_serialization_helper(mangled_name, subtype)?;
+            self.output_deserialization_helper(mangled_name, subtype)?;
+        }
+        self.out.unindent();
+        writeln!(self.out, "}}")?;
+        writeln!(self.out)
+    }
+
+    fn needs_helper(format: &Format) -> bool {
+        use Format::*;
+        match format {
+            Option(_) | Seq(_) | Map { .. } | Tuple(_) | TupleArray { .. } => true,
+            _ => false,
+        }
+    }
+
+    fn quote_serialize_value(&self, value: &str, format: &Format, use_this: bool) -> String {
+        use Format::*;
+        let this_str = if use_this { "this." } else { "" };
+
+        match format {
+            TypeName(_) => format!("{}{}.serialize(serializer);", this_str, value),
+            Unit => format!("serializer.serializeUnit({}{});", this_str, value),
+            Bool => format!("serializer.serializeBool({}{});", this_str, value),
+            I8 => format!("serializer.serializeI8({}{});", this_str, value),
+            I16 => format!("serializer.serializeI16({}{});", this_str, value),
+            I32 => format!("serializer.serializeI32({}{});", this_str, value),
+            I64 => format!("serializer.serializeI64({}{});", this_str, value),
+            I128 => format!("serializer.serializeI128({}{});", this_str, value),
+            U8 => format!("serializer.serializeU8({}{});", this_str, value),
+            U16 => format!("serializer.serializeU16({}{});", this_str, value),
+            U32 => format!("serializer.serializeU32({}{});", this_str, value),
+            U64 => format!("serializer.serializeU64({}{});", this_str, value),
+            U128 => format!("serializer.serializeU128({}{});", this_str, value),
+            F32 => format!("serializer.serializeF32({}{});", this_str, value),
+            F64 => format!("serializer.serializeF64({}{});", this_str, value),
+            Char => format!("serializer.serializeChar({}{});", this_str, value),
+            Str => format!("serializer.serializeStr({}{});", this_str, value),
+            Bytes => format!("serializer.serializeBytes({}{});", this_str, value),
+            _ => format!(
+                "Helpers.serialize{}({}{}, serializer);",
```

Comment:
Just looking at the code, it would be nice to add an underscore here: "Helpers.serialize_{}"

Reply:
I went with the language most common notation (which is serializeType)